src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.Utf8.cs

```diff
@@ -0,0 +1,332 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using System.Text;
+using System.Text.Unicode;
+
+namespace System.Globalization
+{
+    internal static partial class Ordinal
+    {
+        internal static int CompareStringIgnoreCaseUtf8(ref byte strA, int lengthA, ref byte strB, int lengthB)
+        {
+            int length = Math.Min(lengthA, lengthB);
+            int range = length;
+
+            ref byte charA = ref strA;
+            ref byte charB = ref strB;
+
+            const byte maxChar = 0x7F;
+
+            while ((length != 0) && (charA <= maxChar) && (charB <= maxChar))
+            {
+                // Ordinal equals or lowercase equals if the result ends up in the a-z range
+                if (charA == charB ||
+                    ((charA | 0x20) == (charB | 0x20) && char.IsAsciiLetter((char)charA)))
+                {
+                    length--;
+                    charA = ref Unsafe.Add(ref charA, 1);
+                    charB = ref Unsafe.Add(ref charB, 1);
+                }
+                else
+                {
+                    int currentA = charA;
+                    int currentB = charB;
+
+                    // Uppercase both chars if needed
+                    if (char.IsAsciiLetterLower((char)charA))
+                    {
+                        currentA -= 0x20;
+                    }
+                    if (char.IsAsciiLetterLower((char)charB))
+                    {
+                        currentB -= 0x20;
+                    }
+
+                    // Return the (case-insensitive) difference between them.
+                    return currentA - currentB;
+                }
+            }
+
+            if (length == 0)
+            {
+                return lengthA - lengthB;
+            }
+
+            range -= length;
+
+            return CompareStringIgnoreCaseNonAsciiUtf8(ref charA, lengthA - range, ref charB, lengthB - range);
+        }
+
+        internal static int CompareStringIgnoreCaseNonAsciiUtf8(ref byte strA, int lengthA, ref byte strB, int lengthB)
+        {
+            // NLS/ICU doesn't provide native UTF-8 support so we need to convert to UTF-16 and compare that way
+
+            // Convert strA using stackalloc for <= 256 characters and ArrayPool otherwise
+
+            char[]? strAUtf16Array;
+            scoped Span<char> strAUtf16;
+            int strAMaxCharCount = Encoding.UTF8.GetMaxCharCount(lengthA);
```

Comment:
If `lengthA` is `int.MaxValue`, the call to `GetMaxCharCount` will throw `OverflowException`. I don't know if that's something you really want to guard against since it's such an extreme edge case, but thought you should know.